constraints_lang.md

```diff
@@ -0,0 +1,311 @@
+p3 = screen_point(x: 21, y: 23)
+p4 = screen_point(x: 22, y: 23)
+l2 = line(p3, p4)
+// Constrain the lines to be parallel.
+parallel(l1, l2)
+// Add a line segment connecting the others.
+line(p2, p3)
+```
+
+In the standard library, we define `parallel`.
+
+```
+// Constrain two line parameters to be parallel.
+constraint parallel(l1, l2) {
+  v1 = vector_from_line(l1)
+  v2 = vector_from_line(l2)
+  vector_parallel(v1, v2) or vector_antiparallel(v1, v2)
+}
+
+// Return a 3D vector from a line.
+fn vector_from_line(line) {
+  return end_point(line) - start_point(line)
+}
+
+// Constrain two vectors to be parallel.
+constraint vector_parallel(v1, v2) {
+  dot(v1, v2) / (||v1|| * ||v2||) = 1
+}
+
+// Constrain two vectors to be antiparallel.
+constraint vector_antiparallel(v1, v2) {
+  dot(v1, v2) / (||v1|| * ||v2||) = -1
+}
+
+// Return the dot product between two 3D vectors.
+fn dot(v1, v2) {
+  return v1.x * v2.x + v1.y * v2.y + v1.z * v2.z
+}
+```
+
+We introduced a new keyword `constraint` to indicate the effect of applying the function. `fn` is for defining pure, computational functions.
+
+However, functions defined with `constraint` work a bit differently. When a user applies a `constraint`, instead of computing a value and returning it, the function applies constraints to its arguments. From the perspective of the caller, this may have the effect of mutating its input arguments, possibly filling in unknown values. Applications of `constraint` functions do not have a return value. They are more like rules in a logic programming language.
+
+> Aside: In functional programming, we do not call functions; we apply them. When you write down a function, you're creating an abstraction. It's a template that is parameterized. When you fill in the parameter, you're applying the abstraction to specifics. On the other hand, "call" is something you do when you execute imperative commands. You call a subroutine and return from it.
+
+The hope is that when a user applies `constraint`s to arguments, the arguments' unconstrained, and therefore unknown, values are filled in by the constraints.
+
+In the case of `parallel(l1, l2)`, each line argument is made up of two points, each with three components, one for each dimension. So one way to think about it is that the caller is passing in two `struct`s with 6 unknowns each. And these unknowns may be filled in by the `constraint` application.
+
+The points could also be partially constrained. I.e. the `constraint` may partially constrain values by adding relationships, but those relationships may not yet lead to fully constrained points. Additional constraints added in the future may fully constrain them.
+
+## Discussion
+
+### Using Constraints
+
+When you want to construct geometry or define your own constraint functions, you write down a series of equations. These equations take the form:
+
+```
+<expression> <relational operator> <expression>
+```
+
+For example:
+
+```
+p1.x = 5
```

Comment:
I'm not sure I understand this notation; I've only seen `point()`, etc. So this constraint would leave the point partially-constrained?

Reply:
Correct. It's saying that the X-component of the point must be equal to something.